pkg/util/winutil/eventlog/subscription/subscription.go

```diff
@@ -0,0 +1,455 @@
+	if err != nil {
+		return err
+	}
+	hSub, err := q.eventLogAPI.EvtSubscribe(
+		hSubWait,
+		q.channelPath,
+		q.query,
+		bookmarkHandle,
+		q.subscribeOriginFlag|q.subscribeFlags)
+	if err != nil {
+		safeCloseNullHandle(windows.Handle(hSubWait))
+		return err
+	}
+
+	// alloc reusable storage for EvtNext output
+	q.evtNextStorage = make([]evtapi.EventRecordHandle, q.eventBatchCount)
+
+	q.waitEventHandle = hSubWait
+	q.subscriptionHandle = hSub
+
+	if q.useNotifyChannel {
+		hStopWait, err := newStopWaitEvent()
+		if err != nil {
+			evtapi.EvtCloseResultSet(q.eventLogAPI, q.subscriptionHandle)
+			safeCloseNullHandle(windows.Handle(hSubWait))
+			return err
+		}
+		q.stopEventHandle = hStopWait
+
+		// Query loop management
+		q.notifyStop = make(chan struct{})
+		q.notifyNoMoreItems = make(chan struct{})
+		q.notifyNoMoreItemsComplete = make(chan struct{})
+		q.notifyEventsAvailable = make(chan struct{})
+		q.notifyEventsAvailableLoopDone = make(chan struct{})
+
+		// start goroutine to query events for channel
+		q.notifyEventsAvailableWaiter.Add(1)
+		go q.notifyEventsAvailableLoop()
+	}
+
+	q.started = true
+
+	return nil
+}
+
+func (q *pullSubscription) Stop() {
+	if !q.started {
+		return
+	}
+
+	if q.useNotifyChannel {
+		// Signal notifyEventsAvailableLoop to stop
+		windows.SetEvent(windows.Handle(q.stopEventHandle))
+		close(q.notifyStop)
+		// Wait for notifyEventsAvailableLoop to stop
+		q.notifyEventsAvailableWaiter.Wait()
+
+		close(q.notifyNoMoreItems)
+		close(q.notifyNoMoreItemsComplete)
+		safeCloseNullHandle(windows.Handle(q.stopEventHandle))
+	}
+
+	// Cleanup Windows API
+	evtapi.EvtCloseResultSet(q.eventLogAPI, q.subscriptionHandle)
+	safeCloseNullHandle(windows.Handle(q.waitEventHandle))
+
+	q.started = false
+}
+
+// notifyEventsAvailableLoop writes to the notifyEventsAvailable channel
+// when the Windows Event Log API Subscription sets the waitEventHandle.
+// On return, closes the notifyEventsAvailable channel to notify the user
+// of an error or a Stop().
+func (q *pullSubscription) notifyEventsAvailableLoop() {
+	// q.Stop() waits on this goroutine to finish, notify it that we are done
+	defer q.notifyEventsAvailableWaiter.Done()
+	// close the notify channel so the user knows this loop is dead
+	defer close(q.notifyEventsAvailable)
+	// close so internal functions know this loop is dead
+	defer close(q.notifyEventsAvailableLoopDone)
+
+	waiters := []windows.Handle{windows.Handle(q.waitEventHandle), windows.Handle(q.stopEventHandle)}
+
+	for {
+		dwWait, err := windows.WaitForMultipleObjects(waiters, false, windows.INFINITE)
+		if err != nil {
+			// WAIT_FAILED
+			pkglog.Errorf("WaitForMultipleObjects failed: %v", err)
+			return
+		}
+		if dwWait == windows.WAIT_OBJECT_0 {
+			// Event records are available, notify the user
+			pkglog.Debugf("Events are available")
+			select {
+			case <-q.notifyStop:
+				return
+			case q.notifyEventsAvailable <- struct{}{}:
+				break
```

Comment:
what notifyEventsAvailable expression means? that we already reported items availbale and the notification item is still in the channel and prevent double notification and consequently we switch to waiting that reader is done via notifyNoMoreItems?

Reply:
`notifyEventsAvailable` is the channel returned to the user by `EventsAvailable`, which the user is waiting on to know when to call `GetEvents`.

```
	// Notify user that event records are available
	notifyEventsAvailable chan struct{}
```

> prevent double notification and consequently we switch to waiting that reader is done via notifyNoMoreItems?

that's right